core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -598,14 +599,24 @@ private boolean applyFailedShard(RoutingAllocation allocation, ShardRouting fail
                     while (initializingNode.hasNext()) {
                         ShardRouting shardRouting = initializingNode.next();
                         if (shardRouting.isRelocationTargetOf(failedShard)) {
-                            logger.trace("{} is removed due to the failure of the source shard", shardRouting);
-                            initializingNode.remove();
+                            if (failedShard.primary()) {
+                                logger.trace("{} is removed due to the failure of the source shard", shardRouting);
+                                initializingNode.remove();
+                            } else {
+                                logger.trace("{}, relocation source failed, mark as initializing without relocation source", shardRouting);
+                                removedRelocationSourceOnRelocationTarget = true;
+                                initializingNode.removeRelocationSource();
+                            }
+                            break;
                         }
                     }
                 }
             }
-
-            matchedNode.moveToUnassigned(unassignedInfo);
+            if (removedRelocationSourceOnRelocationTarget) {
```

Comment:
I was a bit hesitant on refactoring the existing flow of the method. I can give that a try though if you think it helps.